pkg/pushsync/pushsync.go

```diff
@@ -480,138 +437,90 @@ func (ps *PushSync) measurePushPeer(t time.Time, err error, origin bool) {
 	ps.metrics.PushToPeerTime.WithLabelValues(status).Observe(time.Since(t).Seconds())
 }
 
-func (ps *PushSync) pushPeer(ctx context.Context, skip *skippeers.List, resultChan chan<- receiptResult, doneChan <-chan struct{}, peer swarm.Address, ch swarm.Chunk, origin bool) {
+func (ps *PushSync) prepareCredit(ctx context.Context, peer swarm.Address, ch swarm.Chunk, origin bool) (accounting.Action, error) {
 
-	var (
-		err     error
-		receipt pb.Receipt
-		pushed  bool
-		now     = time.Now()
-	)
-
-	defer func() {
-		select {
-		case resultChan <- receiptResult{pushTime: now, peer: peer, err: err, sent: pushed, receipt: &receipt}:
-		case <-doneChan:
-		}
-	}()
-
-	// compute the price we pay for this receipt and reserve it for the rest of this function
 	receiptPrice := ps.pricer.PeerPrice(peer, ch.Address())
 
-	creditCtx, cancel := context.WithTimeout(context.Background(), time.Second)
+	creditCtx, cancel := context.WithTimeout(ctx, time.Second)
 	defer cancel()
 
-	// Reserve to see whether we can make the request
 	creditAction, err := ps.accounting.PrepareCredit(creditCtx, peer, receiptPrice, origin)
 	if err != nil {
-		skip.Add(ch.Address(), peer, overDraftRefresh)
-		return
-	}
-	defer creditAction.Cleanup()
-
-	skip.Add(ch.Address(), peer, skippeers.MaxDuration)
-
-	stamp, err := ch.Stamp().MarshalBinary()
-	if err != nil {
-		return
+		ps.skipList.Add(ch.Address(), peer, overDraftRefresh)
+		return nil, err
 	}
+	ps.skipList.Add(ch.Address(), peer, sanctionWait)
 
-	streamer, err := ps.streamer.NewStream(ctx, peer, nil, protocolName, protocolVersion, streamName)
-	if err != nil {
-		err = fmt.Errorf("new stream for peer %s: %w", peer, err)
-		return
-	}
-	defer streamer.Close()
+	return creditAction, nil
+}
 
-	w, r := protobuf.NewWriterAndReader(streamer)
-	err = w.WriteMsgWithContext(ctx, &pb.Delivery{
-		Address: ch.Address().Bytes(),
-		Data:    ch.Data(),
-		Stamp:   stamp,
-	})
-	if err != nil {
-		_ = streamer.Reset()
-		err = fmt.Errorf("chunk %s deliver to peer %s: %w", ch.Address(), peer, err)
-		return
-	}
+func (ps *PushSync) push(ctx context.Context, resultChan chan<- receiptResult, doneChan <-chan struct{}, peer swarm.Address, ch swarm.Chunk, action accounting.Action) {
 
-	ps.metrics.TotalSent.Inc()
+	ctx, cancel := context.WithTimeout(ctx, defaultTTL)
+	defer cancel()
 
-	pushed = true
+	var (
+		err     error
+		receipt *pb.Receipt
+		now     = time.Now()
+	)
 
-	// if you manage to get a tag, just increment the respective counter
-	t, err := ps.tagger.Get(ch.TagID())
-	if err == nil && t != nil {
-		err = t.Inc(tags.StateSent)
-		if err != nil {
-			err = fmt.Errorf("tag %d increment: %w", ch.TagID(), err)
-			return
+	defer func() {
+		select {
+		case resultChan <- receiptResult{pushTime: now, peer: peer, err: err, receipt: receipt}:
+		case <-doneChan:
 		}
-	}
+	}()
```

Comment:
could code withing this deferred anonymous function be moved at the end of function? 
this would 1) avoid defer func construct; and 2) var block from 462-465 could be removed?